File: data_structure/js/doublylinkedList.js
//---------------------------------------------------------------
// DOUBLY LINKED LIST
//---------------------------------------------------------------
//
// Scope: the DOUBLY linked list (each node knows both neighbours).
//        See linkedlist.js for the singly linked version.
//
//     null <- [1] <-> [10] <-> [5] -> null
//              ^                ^
//             head             tail
//
// The extra `prev` pointer costs one reference per node and buys:
//   - backwards traversal
//   - O(1) removal of a node you already hold (no walking the list to
//     find its predecessor)
// That second property is why an LRU cache is a hash map plus a
// doubly linked list.
//
// Time  : prepend / append        O(1)
//         removeNode(node)        O(1)
//         get / insert / remove by index  O(N)
// Space : O(N)

class Node {
  constructor(value) {
    this.value = value;
    this.prev = null;
    this.next = null;
  }
}

class DoublyLinkedList {
  // `new DoublyLinkedList(10)` starts with one node; no argument is empty.
  constructor(value) {
    this.head = null;
    this.tail = null;
    this.length = 0;
    if (value !== undefined) this.append(value);
  }

  isEmpty() {
    return this.length === 0;
  }

  prepend(value) {
    const node = new Node(value);
    if (this.head === null) {
      this.head = node;
      this.tail = node;
    } else {
      node.next = this.head;
      this.head.prev = node;
      this.head = node;
    }
    this.length++;
    return node;
  }

  append(value) {
    const node = new Node(value);
    if (this.tail === null) {
      this.head = node;
      this.tail = node;
    } else {
      node.prev = this.tail;
      this.tail.next = node;
      this.tail = node;
    }
    this.length++;
    return node;
  }

  _nodeAt(index) {
    let node = this.head;
    for (let i = 0; i < index; i++) node = node.next;
    return node;
  }

  get(index) {
    if (index < 0 || index >= this.length) return undefined;
    return this._nodeAt(index).value;
  }

  // Insert `value` so that it ends up AT `index`.
  // FOUR pointers get rewired -- draw it before you code it:
  //     leader <-> follower          becomes
  //     leader <-> new <-> follower
  insert(index, value) {
    if (index <= 0) return this.prepend(value);
    if (index >= this.length) return this.append(value);

    const node = new Node(value);
    const leader = this._nodeAt(index - 1);
    const follower = leader.next;

    node.prev = leader;
    node.next = follower;
    leader.next = node;
    follower.prev = node;
    this.length++;
    return node;
  }

  // Unlink a node we already hold -- O(1), no traversal.
  removeNode(node) {
    if (node.prev) node.prev.next = node.next;
    else this.head = node.next; // node was the head

    if (node.next) node.next.prev = node.prev;
    else this.tail = node.prev; // node was the tail

    node.prev = null;
    node.next = null;
    this.length--;
    return node.value;
  }

  // Remove the node at `index` and return its value.
  remove(index) {
    if (index < 0 || index >= this.length) return undefined;
    return this.removeNode(this._nodeAt(index));
  }

  toArray() {
    const values = [];
    for (let node = this.head; node !== null; node = node.next) {
      values.push(node.value);
    }
    return values;
  }

  // Walk backwards from the tail -- impossible in a singly linked list.
  toArrayReversed() {
    const values = [];
    for (let node = this.tail; node !== null; node = node.prev) {
      values.push(node.value);
    }
    return values;
  }
}

// demo
const list = new DoublyLinkedList(10);
list.append(5);
list.append(16);
list.prepend(1);
console.assert(list.toArray().join(',') === '1,10,5,16', 'prepend + append');
console.assert(list.toArrayReversed().join(',') === '16,5,10,1', 'same list, read backwards');

list.insert(2, 99);
console.assert(list.toArray().join(',') === '1,10,99,5,16', 'insert in the middle');
console.assert(list.toArrayReversed().join(',') === '16,5,99,10,1', 'prev links stayed correct');

list.insert(20, 88); // index past the end -> appended
console.assert(list.tail.value === 88 && list.tail.prev.value === 16, 'tail wired both ways');

console.assert(list.get(2) === 99, 'get by index');
console.assert(list.remove(2) === 99, 'remove returns the removed value');
console.assert(list.toArray().join(',') === '1,10,5,16,88');

// O(1) removal of a node we kept a reference to
const node = list.insert(2, 77);
console.assert(list.toArray().join(',') === '1,10,77,5,16,88');
list.removeNode(node);
console.assert(list.toArray().join(',') === '1,10,5,16,88', 'node unlinked without a walk');

// removing the ends keeps head/tail consistent in BOTH directions
list.remove(0);
list.remove(list.length - 1);
console.assert(list.toArray().join(',') === '10,5,16');
console.assert(list.toArrayReversed().join(',') === '16,5,10');
console.assert(list.head.prev === null && list.tail.next === null, 'ends are terminated');

const single = new DoublyLinkedList(1);
single.remove(0);
console.assert(single.isEmpty() && single.head === null && single.tail === null,
  'emptying resets both pointers');

console.log('Success.');

module.exports = { DoublyLinkedList, Node };

File: data_structure/python/doublylinkedlist.py
#---------------------------------------------------------------
# DOUBLY LINKED LIST
#---------------------------------------------------------------
#
# Scope: the DOUBLY linked list (each node knows both neighbours).
#        See linkedList.py for the singly linked version.
#
#     None <- [1] <-> [2] <-> [3] -> None
#              ^                ^
#             head             tail
#
# The extra `prev` pointer costs one reference per node and buys:
#   - backwards traversal
#   - O(1) removal of a node you already hold (no need to walk the
#     list to find its predecessor)
# That second property is exactly why an LRU cache is a hash map
# plus a doubly linked list -- see algorithm/python/lru_cache.py.
#
# Time  : prepend / append          O(1)
#         remove_node(node)         O(1)
#         get / insert / remove by index   O(N)
# Space : O(N)
#
# References:
#   - https://www.geeksforgeeks.org/doubly-linked-list/


class Node:
    """A single link, pointing at BOTH neighbours."""

    def __init__(self, value):
        self.value = value
        self.prev = None
        self.next = None


class DoublyLinkedList:
    """Doubly linked list with head and tail pointers."""

    def __init__(self, values=()):
        self.head = None
        self.tail = None
        self._size = 0
        for value in values:
            self.append(value)

    def __len__(self):
        return self._size

    def __iter__(self):
        node = self.head
        while node:
            yield node.value
            node = node.next

    def __str__(self):
        return " <-> ".join(str(v) for v in self)

    def to_list(self):
        return list(self)

    def to_list_reversed(self):
        """Walk backwards from the tail -- impossible in a singly linked list."""
        values, node = [], self.tail
        while node:
            values.append(node.value)
            node = node.prev
        return values

    def is_empty(self):
        return self.head is None

    def prepend(self, value):
        """Insert at the FRONT."""
        node = Node(value)
        if self.head is None:
            self.head = self.tail = node
        else:
            node.next = self.head
            self.head.prev = node
            self.head = node
        self._size += 1
        return node

    def append(self, value):
        """Insert at the END."""
        node = Node(value)
        if self.tail is None:
            self.head = self.tail = node
        else:
            node.prev = self.tail
            self.tail.next = node
            self.tail = node
        self._size += 1
        return node

    def _node_at(self, index):
        node = self.head
        for _ in range(index):
            node = node.next
        return node

    def get(self, index):
        if not 0 <= index < self._size:
            raise IndexError("index out of range: {}".format(index))
        return self._node_at(index).value

    def insert(self, index, value):
        """Insert `value` so that it ends up AT `index`.

        Four pointers have to be rewired -- draw it before you code it:

            leader <-> follower          becomes
            leader <-> new <-> follower
        """
        if not 0 <= index <= self._size:
            raise IndexError("index out of range: {}".format(index))
        if index == 0:
            return self.prepend(value)
        if index == self._size:
            return self.append(value)

        node = Node(value)
        leader = self._node_at(index - 1)
        follower = leader.next

        node.prev = leader
        node.next = follower
        leader.next = node
        follower.prev = node
        self._size += 1
        return node

    def remove_node(self, node):
        """Unlink a node we already hold -- O(1), no traversal needed."""
        if node.prev:
            node.prev.next = node.next
        else:                                  # node was the head
            self.head = node.next
        if node.next:
            node.next.prev = node.prev
        else:                                  # node was the tail
            self.tail = node.prev
        node.prev = node.next = None
        self._size -= 1
        return node.value

    def remove(self, index):
        """Remove and return the value at `index`."""
        if not 0 <= index < self._size:
            raise IndexError("index out of range: {}".format(index))
        return self.remove_node(self._node_at(index))


if __name__ == "__main__":
    dll = DoublyLinkedList([1, 7, 6])
    assert str(dll) == "1 <-> 7 <-> 6"

    dll.append(4)
    dll.prepend(0)
    assert dll.to_list() == [0, 1, 7, 6, 4]

    # the same list read from the tail backwards
    assert dll.to_list_reversed() == [4, 6, 7, 1, 0]

    dll.insert(3, 8)
    assert dll.to_list() == [0, 1, 7, 8, 6, 4]
    assert dll.to_list_reversed() == [4, 6, 8, 7, 1, 0]

    assert dll.get(3) == 8
    assert dll.remove(3) == 8
    assert dll.to_list() == [0, 1, 7, 6, 4]

    # O(1) removal of a node we kept a reference to
    node = dll.insert(2, 99)
    assert dll.to_list() == [0, 1, 99, 7, 6, 4]
    dll.remove_node(node)
    assert dll.to_list() == [0, 1, 7, 6, 4]

    # removing the ends keeps head/tail consistent in both directions
    dll.remove(0)
    dll.remove(len(dll) - 1)
    assert dll.to_list() == [1, 7, 6]
    assert dll.to_list_reversed() == [6, 7, 1]

    single = DoublyLinkedList([1])
    single.remove(0)
    assert single.is_empty() and single.head is None and single.tail is None

    print("Success.")
